crates/lib/src/execution_data.rs

```diff
@@ -1,120 +1,345 @@
+use num::complex::Complex64;
+use serde::{Deserialize, Serialize};
+use std::collections::{BTreeMap, HashMap};
+use std::num::ParseIntError;
 use std::str::FromStr;
 use std::time::Duration;
 
-use qcs_api_client_grpc::models::controller::ReadoutValues;
-use quil_rs::instruction::MemoryReference;
+use ndarray::prelude::*;
 
-use crate::RegisterData;
+use crate::qpu::readout_data::ReadoutValues;
+use crate::{qpu::readout_data::QpuReadout, qvm::QvmMemory, RegisterData};
 
-/// The result of executing an [`Executable`](crate::Executable) via
-/// [`Executable::execute_on_qvm`](crate::Executable::execute_on_qvm).
+/// Represents the two possible types of readout data returned from either the QVM or a real QPU.
+/// Each variant contains the original data returned from it's respective executor.
+///
+/// # Usage
+///
+/// Your usage of [`ReadoutData`] will depend on the types of programs you are running and where.
+/// The `to_readout_map()` method will attempt to build a [`ReadoutMap`] out of the data, where each
+/// register name is mapped to a 2-dimensional rectangular [`RegisterMatrix`] where each row
+/// represents the final values in each register index for a particular shot. This is often the
+/// desired form of the data and it is _probably_ what you want. This transformation isn't always
+/// possible, in which case `to_readout_map()` will return an error.
+///
+/// To understand why this transformation can fail, we need to understand a bit about how readout data is
+/// returned from the QVM and from a real QPU:
+///
+/// The QVM treats each `DECLARE` statement as initialzing some amount of memory. This memory works
+/// as one might expect it to. It is zero-initalized, and subsequent writes to the same region
+/// overwrite the previous value. The QVM returns memory at the end of every shot. This means
+/// we get the last value in every memory reference for each shot, which is exactly the
+/// representation we want for a [`RegisterMatrix`]. For this reason, `to_readout_map()` should
+/// always succeed for [`ReadoutData::Qvm`]
+///
+/// The QPU on the other hand doesn't use the same memory model as the QVM. Each memory reference
+/// (ie. "ro\[0\]") is more like a stream than a value in memory. Every `MEASURE` to a memory
+/// reference emits a new value to said stream. This means that the number of values per memory
+/// reference can vary per shot. For this reason, it's not always clear what the final value in
+/// each shot was for a particular reference. When this is the case, `to_readout_map()` will return
+/// an error as it's impossible to build a correct [`RegisterMatrix`]  from the data without
+/// knowing the intent of the program that was run. Instead, it's recommended to build the
+/// [`RegisterMatrix`] you need from the inner [`QpuReadout`] data using the knowledge of your
+/// program to choose the correct readout values for each shot.
+#[derive(Debug, Clone, PartialEq, EnumAsInner)]
+pub enum ReadoutData {
```

Comment:
Yep. This is what I had in mind from the previous comment - this might better be `ResultData`, which can wrap:

* QVM memory
* QPU memory _and_ readout data

and then, like you have it here, present an interface for accessing those things in a common way.

Reply:
> QPU memory _and_ readout data

Would you suggest renaming `QpuReadout` if it is supposed to represent memory as well?